### DISCUSSION TITLE
Use CIDR range in ALLOWED_HOSTS

### DISCUSSION THREAD

**Discussion by llamafilm** [netbox-community/netbox#14697] [General]
I've put a load balancer in front of Netbox, with TLS terminated on the load balancer.  To make this work I use these settings:
```
ALLOWED_HOSTS = ['*']
CSRF_TRUSTED_ORIGINS = ['https://netbox.mydomain.net']
```

The load balancer does periodic healthchecks on Netbox's HTTP endpoint, so I would prefer using a CIDR range here like `100.64.0.0/10`.  This doesn't work, it returns 400 error.  I don't think it's a big deal, since I restrict HTTP ingress to only the load balancer, but would it be better to allow CIDR ranges here, and perhaps also partial wildcards like `*.mydomain.net` ?

**Comment by candlerb** [comment]
[ALLOWED_HOSTS](https://docs.netbox.dev/en/stable/configuration/required-parameters/#allowed_hosts) contains the list of hostnames that the Netbox server itself is known as to the outside world. If you put an IP address here, then it's the IP address of the server itself, not the IP address of the load balancer or client.

For example, if your server is running on address 10.11.12.13, and you want people to be able to access it as `http://10.11.12.13` then you'd put `10.11.12.13` in ALLOWED_HOSTS. However, if clients only access it using DNS name, which is normal given that they'll probably expect to see a certificate too, then you can just put the hostname(s) there.

If the loadbalancer is making healthchecks then you ought to be able to tell it which hostname to use (since any web server could be running multiple virtual hosts anyway, and you'll need to specify which one to check)

Therefore, I don't really see your use case for a CIDR range. Are you saying that the IP address of the Netbox server itself changes dynamically? Or you have a pool of Netbox front-end servers? Even then, the loadbalancer should be setting a Host: header so that it's testing `http://netbox.mydomain.net`, not `http://10.11.12.13`

**Reply by llamafilm** [reply]
Every time I redeploy after a code change, the Netbox server gets a new random IP in a known subnet.  This is pretty common behavior with Amazon EC2.

**Reply by jeremystretch** [reply]
You'd typically only add the DNS name(s) relevant to the server. The [Django docs](https://docs.djangoproject.com/en/4.2/ref/settings/#allowed-hosts) might offer more context.

**Reply by llamafilm** [reply]
Thanks, I see there wildcards are supported like `.mydomain.net`.  But that doesn't actually help in this situation because the load balancer queries the instance on its IP address.  I'm not aware of a way to make AWS load balancer check a DNS name instead of IP.

**Reply by markkuleinio** [reply]
If you redeploy the EC2, have your redeploy code deploy that new EC2 in the load balancer as well.

I'm using two CloudFormation templates: first one deploys the EC2, the second one deploys the ALB, referencing the EC2. I don't need to redeploy just the EC2. If I destroy and rebuild the EC2, I'll just destroy and redeploy the ALB as well, and update the company-branded internal CNAME to use the new ALB DNS name.

**Reply by markkuleinio** [reply]
And, I have disabled the default Nginx site config (on the NetBox server) and replaced with it just

```
$ cat /etc/nginx/sites-enabled/lbcheck
server {
    listen 80;
    server_name $server_addr;
    location / {
        return 200 'OK\n';
    }
}
```

Since I only have one NetBox EC2, I don't see much value for the health check anyway but ALB checks "/" for status 200 so there it gets that.

--> My ALLOWED_HOSTS is configured with just mynetboxserver.example.com (= the hostname the clients are using), no need for anything special for ALB to work.

**Reply by llamafilm** [reply]
I use Spinnaker to manage deployment pipelines. There’s no need to recreate the ALB because it automatically gets the IP address of the new instance in the target group.

The dummy healthcheck is a clever idea. That would probably be fine for my situation too. I’m just in the habit of  planning ahead for scale, even when it might be unnecessary.

**Reply by markkuleinio** [reply]
And, if you later want to make "real" health checks to the NetBox level and if you still have the normal Nginx/Apache on the NetBox EC2, maybe you can figure out a config in Nginx/Apache that mangles the headers (for the health checks only) so that no special NetBox-level config is required.

**Comment by markkuleinio** [comment]
Someone please tell me again in which cases is setting **CSRF_TRUSTED_ORIGINS** required?

I mean, I have it in some of my older configs, but in the latest 3.7.0 test installations I haven't set it, and everything seems to work just fine. 

I'm running non-TLS NetBox on EC2 behind the ALB (that does TLS offloading):

client -> ALB -> nginx -> gunicorn (NetBox)

**Reply by llamafilm** [reply]
When I tried without that parameter today  got errors. Using v3.7.0. I’m not sure, but it might have something to do with the way your ALB and nginx process the X-forwarded header. See #8665.

**Reply by markkuleinio** [reply]
This is my Nginx config:

```
$ cat /etc/nginx/sites-enabled/netbox
server {
    listen 80;
    server_name mynetboxserver.example.com;

    client_max_body_size 25m;
    proxy_set_header X-Forwarded-Host $http_host;
    # Makes API token allowed IP check work when behind LB:
    proxy_set_header X-Real-IP $proxy_add_x_forwarded_for;
    # LB offloads TLS (we see only http), so we must force https for all links to work:
    proxy_set_header X-Forwarded-Proto https;
    proxy_read_timeout 300;

    location /static/ {
        alias /opt/netbox/netbox/static/;
    }

    location / {
        proxy_pass http://127.0.0.1:8001;
    }

    error_page 502 /502.html;
    location /502.html {
        return 502 '<html>\n<head><title>502 Bad Gateway</title></head>\n<body><center><h1>NetBox is not running due to maintenance, try again later</h1><p>(Status code: 502 Bad Gateway)</p></center></body>\n</html>\n';
    }
}
```

**Reply by candlerb** [reply]
> Someone please tell me again in which cases is setting **CSRF_TRUSTED_ORIGINS** required?

I think it's to do with any page which does a form POST, e.g. edit device, bulk edit device.  If those things work without it, then I'm not sure :-)

Regarding the load balancer problem from @llamafilm: can you get a tcpdump of the actual headers sent by the LB's HTTP healthcheck - when they arrive at Nginx, and after they have been forwarded to the Netbox server?

```
tcpdump -i eth0 -nn -s0 -A tcp port 80
tcpdump -i lo -nn -s0 -A tcp port 8001
```

If the load balancer is sending `Host: 10.11.12.13`, or no `Host:` header at all, and can't be configured to send `Host: netbox.example.com` in its healthcheck, then that's a poor load balancer. You don't really want to do a dummy healthcheck, because if the Netbox server itself has failed you'd want the load balancer to notice. But if you have no choice, you could set up virtualhosts on Nginx so that incoming requests to `netbox.example.com` are forwarded to gunicorn, and anything else is given a dummy 200 response.

**Reply by markkuleinio** [reply]
> You don't really want to do a dummy healthcheck, because if the Netbox server itself has failed you'd want the load balancer to notice.

Sorry, I didn't mention, but in our case the only purpose for using ALB with a single EC2 is to get the TLS certificate automatically renewed by AWS, so I don't have to manage it on the NetBox server with any custom tooling. That's why the real health checks don't have much meaning in our case.

**Reply by markkuleinio** [reply]
> I think it's to do with any page which does a form POST, e.g. edit device, bulk edit device. If those things work without it, then I'm not sure :-)

Yeah no problems, and the links in racks work, so I don't know what was the problem earlier

**Reply by markkuleinio** [reply]
My ALB health checks look like these:

```
GET / HTTP/1.1
Host: 10.200.200.101
Connection: close
User-Agent: ELB-HealthChecker/2.0
Accept-Encoding: gzip, compressed
```
10.200.200.101 = EC2 IP

**Reply by candlerb** [reply]
Ugh. Assuming you can't override the Host header - and it looks like you [can't](https://docs.aws.amazon.com/elasticloadbalancing/latest/application/target-group-health-checks.html) - then on Nginx put the proxying to Netbox inside a virtualhost for netbox.example.com, and outside the virtualhost return a dummy OK page on `/`

**Reply by markkuleinio** [reply]
Yes, see my earlier comment [above](https://github.com/netbox-community/netbox/discussions/14697#discussioncomment-8020284)

**Reply by markkuleinio** [reply]
FWIW, I get these headers for browser traffic:

From ALB to Nginx:
```
Host: mynetboxserver.example.com
X-Forwarded-For: 10.10.10.184
X-Forwarded-Proto: https
X-Forwarded-Port: 443
```

From Nginx (with those configurations) to NetBox:

```
    proxy_set_header X-Real-IP $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Proto https;
```

```
Host: 127.0.0.1:8001
X-Forwarded-Host: mynetboxserver.example.com
X-Real-IP: 10.10.10.184, 10.200.200.33
X-Forwarded-For: 10.10.10.184
X-Forwarded-Proto: https
X-Forwarded-Port: 443
```

where 10.10.10.184 = client IP, 10.200.200.33 = ALB IP

This works also for API token allowed IP lists, and device links in the racks are working.

**Comment by candlerb** [comment]
> > I think it's to do with any page which does a form POST, e.g. edit device, bulk edit device. If those things work without it, then I'm not sure :-)
> 
> Yeah no problems, and the links in racks work, so I don't know what was the problem earlier

Yes you're right: on my system `CSRF_TRUSTED_ORIGINS` is not present at all in configuration.py, so it defaults in settings.py to empty array:

```
./settings.py:CSRF_TRUSTED_ORIGINS = getattr(configuration, 'CSRF_TRUSTED_ORIGINS', [])
```

So I think you're right, this isn't needed at all normally - maybe only for very weird situations where you have a different website which includes a form which POSTs to Netbox. [Django docs](https://docs.djangoproject.com/en/5.0/ref/settings/#csrf-trusted-origins)

**Comment by candlerb** [comment]
> would it be better to allow CIDR ranges here, and perhaps also partial wildcards like `*.mydomain.net` ?

FYI, `ALLOWED_HOSTS` does allow wildcards like `.mydomain.net`. [Django docs ref](https://docs.djangoproject.com/en/5.0/ref/settings/#allowed-hosts)
